tests/data_classes.py

```diff
@@ -0,0 +1,647 @@
+        # "11.2-RELEASE-p6" or "11.2-RELEASE"
+        release = self.release.split('-', 1)
+        if len(release) == 1:
+            self.release = (999, release[0])
+        elif len(release) == 2:
+            self.release = (float(release[0]), release[1])
+        else:
+            # it is of format 11.2-RELEASE-p6
+            self.release = (
+                float(release[0]), int(release[2][1:]), release[1]
+            )
+
+    def normalize_state(self):
+        self.state = 1 if self.state == 'down' else 0
+
+    def normalize_ip4(self):
+        try:
+            ip = tuple(int(c) for c in self.ip4.split('.'))
+        except ValueError:
+            ip = (300, self.ip4)
+
+        self.ip4 = ip
+
+    def normalize_boot(self):
+        self.boot = 0 if self.boot == 'on' else 1
+
+    def normalize_jid(self):
+        if self.jid.isnumeric():
+            self.jid = int(self.jid)
+        elif self.jid == '-':
+            self.jid = 99999999
+
+    def normalize_size_values(self, value):
+        try:
+            return float(value.strip()[:-1]) * (10 ** self.powers[value[-1]])
+        except ValueError:
+            return 0
+
+    def normalize_values(self):
+        # normalize uses
+        for attr in [
+            a for a in dir(self)
+            if not callable(a) and hasattr(self, f'normalize_{a}')
+            and getattr(self, a)
+        ]:
+            getattr(self, f'normalize_{attr}')()
+
+    # Function to use for sorting of RowBase Objects
+    def sort_flag(self, flag):
+        assert hasattr(self, flag)
+
+        value = getattr(self, flag)
+        if not isinstance(value, tuple):
+            value = (value,)
+
+        # Default to sorting to name always
+        value += (self.name,)
+        return value
+
+    # PARSING FUNCTIONS
+    def full_parse(self):
+        # 10 columns - JID, NAME, BOOT, STATE, TYPE, RELEASE,
+        # IP4, IP6, TEMPLATE, BASEJAIL
+        self.jid, self.name, self.boot, self.state, self.type, \
+        self.release, self.ip4, self.ip6, self.template, self.basejail = \
+            self.standard_parse()
+
+    def all_parse(self):
+        # 4 columns - JID, NAME, STATE, RELEASE, IP4
+        self.jid, self.name, self.state, self.release, self.ip4 = \
+            self.standard_parse()
+
+    def releases_only_parse(self):
+        # 1 column - Bases Fetched
+        self.release = self.standard_parse()[0]
+
+    def quick_parse(self):
+        # 2 column - NAME, IP4
+        # In case of releases, it will only be one # NAME
+        data = self.standard_parse()
+        if len(data) == 1:
+            self.release = data[0]
+        else:
+            self.name, self.ip4 = data
+
+    def df_parse(self):
+        self.name, self.crt, self.res, self.qta, self.use, self.ava = \
+            self.standard_parse()
+
+    def snapshot_parse(self):
+        self.name, self.created, self.rsize, self.used = self.standard_parse()
+
+
+class ZFS:
```

Comment:
Perhaps we shift to subprocess zfs again starting here as well? That way we can remove our dep on it sooner and easier.